lib/src/entrypoint.dart

```diff
@@ -122,31 +123,47 @@ class Entrypoint {
 
   PackageGraph _packageGraph;
 
+  /// Where the lock file and package configurations are to be found.
+  ///
+  /// Global packages (except those from path source)
+  /// store these in the global cache.
+  String get _configPath => root.dir == null || isCached
+      ? p.join(cache.rootDir, 'global_packages', root.name)
+      : root.dir;
+
   /// The path to the entrypoint's "packages" directory.
   String get packagesPath => root.path('packages');
 
   /// The path to the entrypoint's ".packages" file.
-  String get packagesFile => root.path('.packages');
+  String get packagesFile => p.join(_configPath, '.packages');
 
   /// The path to the entrypoint's ".dart_tool/package_config.json" file.
   String get packageConfigFile =>
-      root.path('.dart_tool', 'package_config.json');
+      p.join(_configPath, '.dart_tool', 'package_config.json');
 
   /// The path to the entrypoint package's pubspec.
   String get pubspecPath => root.path('pubspec.yaml');
 
   /// The path to the entrypoint package's lockfile.
-  String get lockFilePath => root.path('pubspec.lock');
+  String get lockFilePath => p.join(_configPath, 'pubspec.lock');
 
   /// The path to the entrypoint package's `.dart_tool/pub` cache directory.
   ///
   /// If the old-style `.pub` directory is being used, this returns that
   /// instead.
   String get cachePath {
-    var newPath = root.path('.dart_tool/pub');
-    var oldPath = root.path('.pub');
-    if (!dirExists(newPath) && dirExists(oldPath)) return oldPath;
-    return newPath;
+    if (isGlobal) {
+      return p.join(
+        cache.rootDir,
+        'global_packages',
+        root.name,
+      );
```

Comment:
Why is this not using `_configPath`.

And why is this different for global packages installed with `pub global activate --source path`

Reply:
Documentation comment about why this is different would be nice.